app/models/client_history.rb

```diff
@@ -0,0 +1,147 @@
+###
+# Copyright 2016 - 2024 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+class ClientHistory
+  attr_reader :client, :user, :requesting_user, :years
+  def initialize(
+    client_id:,
+    user_id:,
+    years:
+  )
+
+    @user = User.system_user
+
+    # The user that requested the PDF generation. If job was kicked off from CAS, this is nil.
+    @requesting_user = User.find_by(id: user_id.to_i)
+    @years = years
+
+    @client = ::GrdaWarehouse::Hud::Client.destination.find(client_id.to_i)
+    @dates = set_pdf_dates(client: @client, requesting_user: @requesting_user)
+  end
+
+  # Limit to Residential Homeless programs
+  def dates
+    @dates.transform_values do |data|
+      data.select { |en| ::HudUtility2024.residential_project_type_ids.include?(en[:project_type]) }
+    end
+    @dates
+  end
+
+  def ordered_dates
+    dates.keys.sort
+  end
+
+  def chronic
+    ::GrdaWarehouse::Config.get(:chronic_definition).to_sym == :chronics ? client.potentially_chronic?(on_date: Date.current) : client.hud_chronic?(on_date: Date.current)
+  end
+
+  def organization_counts
+    dates.values.flatten.group_by { |en| HudUtility2024.project_type en[:organization_name] }.transform_values(&:count)
+  end
+
+  def project_type_counts
+    dates.values.flatten.group_by { |en| HudUtility2024.project_type en[:project_type] }.transform_values(&:count)
+  end
+
+  def generate_service_history_pdf
+    template_file = 'client_access_control/history/pdf'
+    layout = false
+    html = PdfGenerator.html(
+      controller: ClientAccessControl::HistoryController,
+      template: template_file,
+      layout: layout,
+      user: user,
+      assigns: {
+        client_history: self,
+      },
+    )
+
+    file_name = 'service_history.pdf'
+    pdf = nil
+    PdfGenerator.new.perform(
+      html: html,
+      file_name: file_name,
+    ) do |io|
+      pdf = io.read
+    end
+
+    client_file = ::GrdaWarehouse::ClientFile.new
+    client_file.client_id = client.id
+    client_file.user_id = requesting_user&.id || user.id
+    client_file.note = "Auto Generated for prior #{years} years"
+    client_file.name = file_name
+    client_file.visible_in_window = true
+    client_file.effective_date = Date.current
+    client_file.tag_list.add(['Homeless Verification'])
```

Comment:
This is fine as is, but might be nicer as part of the new statement.  Any idea why we did it this way?
```suggestion
    client_file = ::GrdaWarehouse::ClientFile.new(
    client_id: client.id,
    user_id: (requesting_user&.id || user.id),
    note: "Auto Generated for prior #{years} years",
    name: file_name,
    visible_in_window: true,
    effective_date: Date.current,
)
    client_file.tag_list.add(['Homeless Verification'])
```

Reply:
This was the format being used in the previous file, and I just copied it forward. I like the cleanup though and will push it into the PR.